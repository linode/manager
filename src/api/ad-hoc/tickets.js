import { fetchFile } from '../fetch';


export function addTicketAttachment(id, attachment) {
  const data = new FormData();
  data.append('file', attachment);
  return fetchFile.post(`/support/tickets/${id}/attachments`, data, null);
}
