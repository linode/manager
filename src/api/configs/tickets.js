import {
  genConfig, ReducerGenerator, genActions, ONE, MANY, POST,
} from '~/api/apiResultActionReducerGenerator';

export const config = genConfig({
  plural: 'tickets',
  endpoint: id => `/support/tickets/${id}`,
  supports: [ONE, MANY, POST],
  subresources: {
    _replies: {
      plural: 'replies',
      endpoint: (ticket, reply) => `/support/tickets/${ticket}/replies/${reply}`,
      supports: [ONE, MANY, POST],
    },
  },
});

export const actions = genActions(config);
export const { reducer } = new ReducerGenerator(config);
