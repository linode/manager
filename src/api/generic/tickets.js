import {
  addParentRefs, ReducerGenerator, genActions, ONE, MANY, POST,
} from '~/api/internal.ts';

export const config = addParentRefs({
  name: 'tickets',
  primaryKey: 'id',
  endpoint: id => `/support/tickets/${id}`,
  supports: [ONE, MANY, POST],
  subresources: {
    _replies: {
      name: 'replies',
      primaryKey: 'id',
      endpoint: (ticket, reply) => `/support/tickets/${ticket}/replies/${reply}`,
      supports: [ONE, MANY, POST],
    },
  },
});

export const actions = genActions(config);
export const { reducer } = new ReducerGenerator(config);
