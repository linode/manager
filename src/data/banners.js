export const testApiBanner = {
  entity: {
    id: 1234,
    type: 'ticket',
    label: 'testing ticket',
  },
  type: 'important_ticket',
  when: null,
};

export const testBillingBanner = {
  ...testApiBanner,
  entity: {
    id: 1235,
    label: 'Please update your billing information.',
  },
};

export const testAbuseBanner = {
  ...testApiBanner,
  entity: {
    id: 1235,
    label: 'ToS Violation',
  },
  type: 'abuse_ticket',
};

export const testAbuseBanner2 = {
  ...testAbuseBanner,
  entity: {
    id: 1236,
  },
};

export const importantBanner = [
  testApiBanner,
];

export const importantBanners = [
  testApiBanner,
  testBillingBanner,
];

export const abuseBanner = [
  testAbuseBanner,
];

export const abuseBanners = [
  testAbuseBanner,
  testAbuseBanner2,
];

export const banners = [
  testApiBanner,
  testBillingBanner,
  testAbuseBanner,
];
