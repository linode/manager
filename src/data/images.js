export const apiTestImage = {
  id: 'linode/arch2016.05',
  recommended: true,
  vendor: 'Arch',
  label: 'Arch Linux 2016.05',
  size: 550,
  created: '2009-01-01T00:00:00',
  description: 'Test Image',
  type: 'manual',
  created_by: 'linode',
  is_public: true,
  deprecated: false,
};

export const testLinodeImage = {
  ...apiTestImage,
  _polling: false,
};

export const publicImages = {
  'linode/arch2016.05': testLinodeImage,
  'linode/debian7': {
    ...testLinodeImage,
    id: 'linode/debian7',
    vendor: 'Debian',
    label: 'Debian 7',
  },
  'linode/debian8.1': {
    ...testLinodeImage,
    id: 'linode/debian8.1',
    vendor: 'Debian',
    label: 'Debian 8.1',
  },
  'linode/debian6': {
    ...testLinodeImage,
    id: 'linode/debian6',
    recommended: false,
    vendor: 'Debian',
    label: 'Debian 6',
  },
};

export const testPrivateImage = {
  ...testLinodeImage,
  id: 'private/38',
  label: 'testImage',
  description: 'image details',
  vendor: null,
  created: '2017-08-08T13:55:16',
  created_by: 'zues',
  size: 2048,
  type: 'manual',
  is_public: false,
};

export const privateImages = {
  'private/38': testPrivateImage,
  'private/39': {
    ...testPrivateImage,
    id: 'private/39',
    label: 'nextImage',
    description: 'image from deleted disk',
    created: '2017-08-08T13:55:16',
    created_by: 'zues',
    size: 2048,
    type: 'automatic',
    is_public: false,
  },
};

export const images = {
  ...privateImages,
  ...publicImages,
};
