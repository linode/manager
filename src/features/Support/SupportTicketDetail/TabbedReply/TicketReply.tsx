import * as React from 'react';
import {
  StyleRulesCallback,
  WithStyles,
  withStyles
} from 'src/components/core/styles';
import TextField from 'src/components/TextField';

type ClassNames = 'replyField';

const styles: StyleRulesCallback<ClassNames> = theme => ({
  replyField: {
    maxHeight: 200,
    marginTop: 0,
    '& > div': {
      maxWidth: '100% !important'
    }
  }
});

export interface Props {
  error?: string;
  handleChange: (value: string) => void;
  value: string;
  placeholder?: string;
}

type CombinedProps = Props & WithStyles<ClassNames>;

class TicketReply extends React.Component<CombinedProps> {
  render() {
    const { placeholder, classes, value, handleChange, error } = this.props;

    return (
      <React.Fragment>
        <TextField
          className={classes.replyField}
          multiline
          rows={9}
          value={value}
          placeholder={placeholder || 'Enter your reply'}
          data-qa-ticket-description
          onChange={e => handleChange(e.target.value)}
          errorText={error}
        />
      </React.Fragment>
    );
  }
}

const styled = withStyles(styles);

export default styled(TicketReply);
