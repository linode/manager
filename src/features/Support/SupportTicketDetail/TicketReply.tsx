import { compose, lensPath, pathOr,  set } from 'ramda';
import * as React from 'react';

import InputAdornment from '@material-ui/core/InputAdornment';
import { StyleRulesCallback, Theme, WithStyles, withStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import AttachFile from '@material-ui/icons/AttachFile';
import CloudUpload from '@material-ui/icons/CloudUpload';

import ActionsPanel from 'src/components/ActionsPanel';
import Button from 'src/components/Button';
import Grid from 'src/components/Grid';
import LinearProgress from 'src/components/LinearProgress';
import Notice from 'src/components/Notice';
import TextField from 'src/components/TextField';
import { createReply, uploadAttachment } from 'src/services/support';
import getAPIErrorFor from 'src/utilities/getAPIErrorFor';

type ClassNames =
  'root'
  | 'form'
  | 'attachFileButton'
  | 'attachmentsContainer'
  | 'attachmentField'
  | 'replyField'
  | 'uploadProgress'

const styles: StyleRulesCallback<ClassNames> = (theme: Theme & Linode.Theme) => ({
  root: {
    width: '100%',
  },
  form: {
    minWidth: '100% !important',
    width: '100vw !important',
  },
  attachFileButton: {
    paddingLeft: 14,
    paddingRight: 20,
  },
  attachmentsContainer: {
    maxWidth: 800,
  },
  attachmentField: {
    marginTop: 0,
    width: 415,
    [theme.breakpoints.down('xs')]: {
      width: 165,
    },
    '& > div ': {
      backgroundColor: theme.bg.main,
      border: 0,
    },
    '& svg': {
      color: theme.palette.text.primary,
      width: 24, 
      fontSize: 22, 
    },
  },
  replyField: {
    paddingBottom: theme.spacing.unit * 4,
    marginBottom: theme.spacing.unit * 2,
    borderBottom: '1px solid #ddd',
    '& > div': {
      maxWidth: '100% !important',
    },
  },
  uploadProgress: {
    maxWidth: 415,
  },
});

interface Props {
  ticketId: number;
  onSuccess: (newReply: Linode.SupportReply) => void;
  reloadAttachments: () => void;
}

interface FileAttachment {
  name: string,
  file: File,
  /* Used to keep track of initial upload status */
  uploading: boolean,
  /* Used to ensure that the file doesn't get uploaded again */
  uploaded: boolean,
  /* Each file needs to keep track of its own errors because each request hits the same endpoint */
  errors?: Linode.ApiFieldError[];
}

interface State {
  value: string;
  submitting: boolean;
  files: FileAttachment[];
  errors?: Linode.ApiFieldError[];
}

type CombinedProps = Props & WithStyles<ClassNames>;

class TicketReply extends React.Component<CombinedProps, State> {
  state: State = {
    value: '',
    submitting: false,
    files: [],
  }

  inputRef = React.createRef<HTMLInputElement>();

  handleReplyInput = (e:React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ value: e.target.value });
  }

  submitForm = () => {
    const { onSuccess, ticketId } = this.props;
    const { value, files } = this.state;

    this.setState({
      submitting: true,
      errors: [],
    });

    /* Send the reply */
    createReply({ description: value, ticket_id: ticketId })
      .then((response) => {
        onSuccess(response.data);
        this.setState({ submitting: false, value: '' });
      })
      .then(() => {
        /* Make sure the reply will go through before attaching files */
        /* Send each file */
        files.map((file, idx) => {
          if (file.uploaded) { return ; } 
          this.setState(set(lensPath(['files', idx, 'uploading']), true));
          const formData = new FormData(); 
          formData.append('file', file.file);
          uploadAttachment(this.props.ticketId, formData)
            .then(() => {
              this.setState(compose(
                /* null out an uploaded file after upload */
                set(lensPath(['files', idx, 'file']), null),
                set(lensPath(['files', idx, 'uploading']), false),
                set(lensPath(['files', idx, 'uploaded']), true),
              ));
              this.props.reloadAttachments();
            })
            /* 
            * Note! We want the first few uploads to succeed even if the last few
            * fail! Don't try to aggregate errors!
            */
            .catch((errors) => {
              this.setState(set(lensPath(['files', idx, 'uploading']), false));
              const error = [{ 'reason': 'There was an error attaching this file. Please try again.' }];
              const newErrors = pathOr(error, ['response', 'data', 'errors'], errors);
              this.setState(set(lensPath(['files', idx, 'errors']), newErrors));
            })
        })
      })
      .catch((errors) => {
        const error = [{ 'reason': 'There was an error creating your reply. Please try again.' }];
        const newErrors = pathOr(error, ['response', 'data', 'errors'], errors);
        this.setState({
          errors: newErrors,
          submitting: false
        });
      })
  }

  clickAttachButton = (e: React.MouseEvent<HTMLButtonElement>) => {
    if (this.inputRef.current) {
      this.inputRef.current.click();
    }
  }

  handleFileSelected = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.target.files && e.target.files.length) {
      this.setState({
        files: [
          ...this.state.files,
          {
            name: e.target.files[0].name,
            /* The file! These can be quite big */
            file: e.target.files[0],
            /* Used to keep track of initial upload status */
            uploading: false,
            /* Used to ensure that the file doesn't get uploaded again */
            uploaded: false,
          }
        ]
      }, () => {
        if (this.inputRef.current) {
          this.inputRef.current.value = '';
        }
      })
    }
  }

  removeFile = (e: React.MouseEvent<HTMLButtonElement>) => {
    if (!e.target) { return; }
    const aidx = e.currentTarget.getAttribute('data-file-idx');
    if (!aidx) { return; }
    const idx = +aidx;
    this.setState({
      files: this.state.files.filter((_, i) => i !== idx),
    }, () => {
      if (this.inputRef.current) {
        this.inputRef.current.value = '';
      }
    });
  }

  render() {
    const { classes } = this.props;
    const { errors, submitting, value, files } = this.state;

    const hasErrorFor = getAPIErrorFor({
      description: 'description',
    }, errors);

    const replyError = hasErrorFor('description');
    const generalError = hasErrorFor('none');

    return (
      <Grid className={classes.root} item>
        <Typography variant="headline" className={classes.root} data-qa-title >
          Reply
        </Typography>
        {generalError && <Notice error spacingBottom={8} spacingTop={16} text={generalError} />}
        <TextField
          className={classes.replyField}
          multiline
          rows={5}
          value={value}
          placeholder="Enter your reply"
          onChange={this.handleReplyInput}
          errorText={replyError}
        />
        {files.map((file, idx) => (
          file.uploaded
            ? null /* this file has already been uploaded so don't show it */
            : (
              <React.Fragment key={idx}>
                <Grid container className={classes.attachmentsContainer}>
                  <Grid item>
                    <TextField
                      className={classes.attachmentField}
                      value={file.name}
                      errorText={file.errors && file.errors.length && file.errors[0].reason}
                      InputProps={{
                        startAdornment:
                        <InputAdornment position="end">
                          <CloudUpload />
                        </InputAdornment>
                      }}
                    />
                    
                  </Grid>
                  <Grid item>
                    <Button
                      type="remove"
                      data-file-idx={idx}
                      onClick={this.removeFile}
                    />
                  </Grid>
                  {file.uploading &&
                    <Grid item xs={12}>
                      <LinearProgress className={classes.uploadProgress} variant="indeterminate"/>
                    </Grid>
                  }
                </Grid>
              </React.Fragment>
            )
        ))}
        <ActionsPanel style={{ marginTop: 16 }}>
          <Button
            type="primary"
            loading={submitting}
            onClick={this.submitForm}
          >
            Send
          </Button>
          <input
            ref={this.inputRef}
            type="file"
            id="attach-file"
            style={{ display: 'none' }}
            onChange={this.handleFileSelected}
          />
          <Button
            component="span"
            className={classes.attachFileButton}
            type="secondary"
            onClick={this.clickAttachButton}
          >
            <AttachFile />
            Attach a file
          </Button>
        </ActionsPanel>
      </Grid>
    )
  }
}

const styled = withStyles(styles, { withTheme: true });

export default styled(TicketReply);
