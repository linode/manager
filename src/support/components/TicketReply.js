import React, { PropTypes } from 'react';

import { Card, CardImageHeader } from 'linode-components/cards';
import TimeDisplay from '~/components/TimeDisplay';

export function getLineBreakCharacter(text) {
  // Check the longer one first because it contains the second one.
  const lineBreakCharacters = ['\r\n', '\n'];
  for (let i = 0; i < lineBreakCharacters.length; i++) {
    const lineBreakCharacter = lineBreakCharacters[i];
    if (text.indexOf(lineBreakCharacter) !== -1) {
      return lineBreakCharacter;
    }
  }

  return null;
}

export function stringToParagraphs(text) {
  const lineBreakCharacter = getLineBreakCharacter(text);
  const ps = text.split(lineBreakCharacter + lineBreakCharacter);

  if (!lineBreakCharacter) {
    return text;
  }

  const elements = [];
  for (let i = 0; i < ps.length; i++) {
    const p = ps[i];

    if (p === '') {
      elements.push(<br key={i} />);
    } else if (p.indexOf('    ') !== -1) {
      return <pre>{text}</pre>;
    } else {
      const linesInP = p.split(lineBreakCharacter).map((line, j) =>
        <span key={j}>{j > 0 ? <br /> : null}{line}</span>);
      elements.push(<p key={i}>{linesInP}</p>);
    }
  }

  return (
    <span>{elements}</span>
  );
}

export default function TicketReply(props) {
  const createdBy = props.reply[`${props.createdField}_by`];
  const created = <TimeDisplay time={props.reply[props.createdField]} />;
  const icon = `https://gravatar.com/avatar/${props.reply.gravatar_id}`;
  const header = (
    <CardImageHeader
      title={createdBy}
      subtitle={created}
      icon={icon}
      tags={props.reply.from_linode ? ['Linode Employee'] : []}
    />
  );

  const description = stringToParagraphs(props.reply.description);

  return (
    <Card header={header} className="TicketReply">
      {description}
    </Card>
  );
}

TicketReply.propTypes = {
  createdField: PropTypes.string.isRequired,
  reply: PropTypes.object.isRequired,
};
