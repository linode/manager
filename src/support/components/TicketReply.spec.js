import React from 'react';
import { mount, shallow } from 'enzyme';

import TicketReply, {
  stringToParagraphs,
  getLineBreakCharacter,
} from '~/support/components/TicketReply';
import { testTicket } from '~/data/tickets';

describe('support/components/TicketReply', () => {
  it('renders all data correctly', () => {
    const page = mount(<TicketReply createdField="opened" reply={testTicket} />);
    const header = page.find('CardImageHeader');
    const time = header.find('TimeDisplay');

    expect(header.props().title).toBe(testTicket.opened_by);
    expect(time.props().time).toBe(testTicket.opened);

    expect(page.find('.Card-body').text()).toBe(
      stringToParagraphs(testTicket.description));
  });

  it('gets the right line break character', () => {
    expect(getLineBreakCharacter('')).toBe(null);
    expect(getLineBreakCharacter('foo\r\nfoo')).toBe('\r\n');
    expect(getLineBreakCharacter('foo\nfoo')).toBe('\n');
  });

  it('doesn\'t format text without newlines', () => {
    const sentence = 'This sentence has no newlines.';
    expect(stringToParagraphs(sentence)).toBe(sentence);
  });

  it('puts string containing four spaces in a pre tag', () => {
    const sentence = 'This sentence\nhas four    spaces.';
    const formatted = shallow(stringToParagraphs(sentence));
    expect(formatted.find('pre').length).toBe(1);
    expect(formatted.find('pre').text()).toBe(sentence);

    expect(formatted.find('br').length).toBe(0);
  });

  it('puts string containing two new lines into a paragraph', () => {
    const firstSentence = 'This sentence is the first paragraph.';
    const secondSentence = 'This sentence is the second.';
    const paragraphs = `${firstSentence}\n\n${secondSentence}`;
    const formatted = shallow(stringToParagraphs(paragraphs));
    expect(formatted.find('p').length).toBe(2);
    expect(formatted.find('p').at(0).text()).toBe(firstSentence);
    expect(formatted.find('p').at(1).text()).toBe(secondSentence);

    expect(formatted.find('br').length).toBe(0);
    expect(formatted.text().indexOf('\n')).toBe(-1);
  });

  it('puts string containing one new line into a paragraph with line breaks', () => {
    const firstSentence = 'This is the first sentence.';
    const secondSentence = 'This sentence is the second.';
    const paragraphs = `${firstSentence}\n${secondSentence}`;
    const formatted = shallow(stringToParagraphs(paragraphs));
    expect(formatted.find('p').length).toBe(1);
    // <br /> replaces the one new line above
    expect(formatted.find('br').length).toBe(1);
    expect(formatted.find('span').at(1).text()).toBe(firstSentence);
    expect(formatted.find('span').at(2).text()).toBe(secondSentence);

    expect(formatted.text().indexOf('\n')).toBe(-1);
  });
});
