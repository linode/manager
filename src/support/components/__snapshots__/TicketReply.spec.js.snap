// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`support/components/TicketReply should render without error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TicketReply
    createdField="opened"
    reply={
        Object {
            "_replies": Object {
              "replies": Array [
                Object {
                  "created": "2017-03-30T16:48:10",
                  "created_by": "support",
                  "description": "This is the first response.",
                  "id": 1,
                },
              ],
            },
            "closed": null,
            "description": "Hullo!",
            "entity": Object {
              "label": "awkward-aimless-alpha-22",
              "type": "linode",
            },
            "id": 1,
            "opened": "2017-03-03T01:22:32",
            "opened_by": "tdude",
            "status": "new",
            "summary": "This is my ticket!",
            "updated": "2017-03-03T01:22:32",
            "updated_by": "tdude",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        "Hullo!",
        null,
      ],
      "className": "TicketReply",
      "header": <CardImageHeader
        icon="https://gravatar.com/avatar/undefined"
        iconClass=""
        subtitle={
                <TimeDisplay
                        time="2017-03-03T01:22:32"
                />
        }
        tags={Array []}
        title="tdude"
/>,
    },
    "ref": null,
    "rendered": Array [
      "Hullo!",
      null,
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          "Hullo!",
          null,
        ],
        "className": "TicketReply",
        "header": <CardImageHeader
          icon="https://gravatar.com/avatar/undefined"
          iconClass=""
          subtitle={
                    <TimeDisplay
                              time="2017-03-03T01:22:32"
                    />
          }
          tags={Array []}
          title="tdude"
/>,
      },
      "ref": null,
      "rendered": Array [
        "Hullo!",
        null,
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
